ip_country/src/country_block_serde.rs

```diff
@@ -154,133 +166,321 @@ impl CountryBlockSerializer {
-                        self.prev_record.start,
-                        minus_one_ipv4(next_record.start),
-                    ),
-                    country: Country::from(self.prev_record.country_idx),
-                };
-                self.prev_record = next_record;
-                Some(prev_block)
-            }
-            None => {
-                self.empty = true;
-                Some(CountryBlock {
-                    ip_range: IpRange::V4(
-                        self.prev_record.start,
-                        Ipv4Addr::new(255, 255, 255, 255),
-                    ),
-                    country: Country::from(self.prev_record.country_idx),
-                })
-            }
-        }
+        self.next_generic()
+    }
+}
+
+impl CountryBlockDeserializer<Ipv6Addr, u16, 8> {
+    pub(crate) fn new(country_data: (Vec<u64>, usize)) -> Self {
+        Self::new_generic(
+            country_data,
+            Ipv6Addr::new(
+                0xFFFF, 0xFFFF, 0xFFFF, 0xFFFF, 0xFFFF, 0xFFFF, 0xFFFF, 0xFFFE,
+            ),
+        )
+    }
+}
+
+impl Iterator for CountryBlockDeserializer<Ipv6Addr, u16, 8> {
+    type Item = CountryBlock;
+
+    fn next(&mut self) -> Option<CountryBlock> {
+        self.next_generic()
+    }
+}
+
+impl DeserializerPrivate<Ipv4Addr> for CountryBlockDeserializer<Ipv4Addr, u8, 4> {
+    fn max_ip_value() -> Ipv4Addr {
+        Ipv4Addr::new(0xFF, 0xFF, 0xFF, 0xFF)
+    }
+
+    fn read_difference_count(bit_queue: &mut BitQueue) -> Option<usize> {
+        Some((bit_queue.take_bits(2)? + 1) as usize)
+    }
+
+    fn read_differences(bit_queue: &mut BitQueue, difference_count: usize) -> Vec<Difference> {
+        Self::read_differences_generic(bit_queue, difference_count, 2, 8)
+    }
+}
+
+impl DeserializerPrivate<Ipv6Addr> for CountryBlockDeserializer<Ipv6Addr, u16, 8> {
+    fn max_ip_value() -> Ipv6Addr {
+        Ipv6Addr::new(
+            0xFFFF, 0xFFFF, 0xFFFF, 0xFFFF, 0xFFFF, 0xFFFF, 0xFFFF, 0xFFFF,
+        )
+    }
+
+    fn read_difference_count(bit_queue: &mut BitQueue) -> Option<usize> {
+        Some((bit_queue.take_bits(3)? + 1) as usize)
+    }
+
+    fn read_differences(bit_queue: &mut BitQueue, difference_count: usize) -> Vec<Difference> {
+        Self::read_differences_generic(bit_queue, difference_count, 3, 16)
     }
 }
 
-impl CountryBlockDeserializerIpv4 {
-    pub fn new(country_data_ipv4: (Vec<u64>, usize)) -> Self {
-        let mut bit_queue = bit_queue_from_country_data(country_data_ipv4);
-        let prev_record = Self::get_record(&mut bit_queue, Ipv4Addr::new(255, 255, 255, 254))
+impl<IPType, SegmentNumRep, const SEGMENTS_COUNT: usize>
+    CountryBlockDeserializer<IPType, SegmentNumRep, SEGMENTS_COUNT>
+where
+    Self: DeserializerPrivate<IPType>,
+    IPType: IPIntoSegments<SegmentNumRep, SEGMENTS_COUNT>
+        + PlusMinusOneIP
+        + From<[SegmentNumRep; SEGMENTS_COUNT]>
+        + Copy
+        + Debug
+        + Display,
```

Comment:
I'm a bit surprised we need both Debug and Display to print an IP. I don't know if you really wanna do that, it may or may not be a hassle, but I'd try to remove the Display and also at other places in the file. I couldn't find where we would've needed it.

Reply:
impl Debug solves this, I could remove that